### PR TITLE
new: add sparse vectors support in discovery and recommend in local mode

### PR DESCRIPTION
**Pull request by joein**
more than a half of the PR are tests 🙈

### CODE REVIEW

**Review by agourlay — APPROVED** [review]
The test coverage looks excellent!

I did not review in details the scoring implementation but I am confident due to the congruence tests :+1:

File: qdrant_client/local/distances.py [hunk 1]
```diff
@@ -49,12 +53,12 @@ def __init__(self, context_pairs: List[ContextPair]):
         self.context_pairs = context_pairs
 
 
-QueryVector = Union[DiscoveryQuery, ContextQuery, RecoQuery, types.NumpyArray, SparseVector]
-
-
-class DistanceOrder(str, Enum):
-    BIGGER_IS_BETTER = "bigger_is_better"
-    SMALLER_IS_BETTER = "smaller_is_better"
+QueryVector = Union[
```
Comment (coszio) [inline qdrant_client/local/distances.py:56]: Only a suggestion, I'll leave it to your criteria
```suggestion
DenseQueryVector = Union[
```

File: qdrant_client/local/local_collection.py [hunk 2]
```diff
@@ -588,7 +634,28 @@ def _preprocess_recommend(
...
         lookup_from_vector_name: Optional[str] = None,
-    ) -> Tuple[List[List[float]], List[List[float]], types.Filter]:
+    ) -> Tuple[
+        List[List[float]],
+        List[List[float]],
+        List[models.SparseVector],
+        List[models.SparseVector],
+        types.Filter,
+    ]:
+        def split_examples(
```
Comment (coszio) [inline qdrant_client/local/local_collection.py:644]: I had to read a few times to understand what this was doing, maybe:
```suggestion
        def examples_into_vectors(
```
?

File: qdrant_client/local/local_collection.py [hunk 3]
```diff
@@ -702,20 +772,61 @@ def recommend(
...
+                    positive_vectors,
+                    negative_vectors,
+                )
+            elif sparse_positive_vectors:
+                query_vector = self._recommend_average_sparse(
+                    sparse_positive_vectors,
+                    sparse_negative_vectors,
+                )
+            else:
+                raise ValueError("No positive examples given")
```
Comment (coszio) [inline qdrant_client/local/local_collection.py:787]:
```suggestion
                raise ValueError("No positive examples given with 'average_vector' strategy")
```

File: qdrant_client/local/local_collection.py [hunk 4]
```diff
@@ -702,20 +772,61 @@ def recommend(
...
-            )
+            if positive_vectors or negative_vectors:
+                query_vector = RecoQuery(
+                    positive=positive_vectors,
+                    negative=negative_vectors,
+                )
+            else:
+                query_vector = SparseRecoQuery(
+                    positive=sparse_positive_vectors, negative=sparse_negative_vectors
+                )
```
Comment (coszio) [inline qdrant_client/local/local_collection.py:798]: ~what if there are no `sparse_positive_vectors` nor `sparse_negative_vectors` either?~
Edit: it is already checked in https://github.com/qdrant/qdrant-client/blob/164c8375334510279e38cc1ca4912ac892705ab7/qdrant_client/local/local_collection.py#L676

File: qdrant_client/local/local_collection.py [hunk 5]
```diff
@@ -848,25 +931,71 @@ def _preprocess_discover(
...
+                dense_context_vectors.append(
+                    ContextPair(positive=pair_vectors[0], negative=pair_vectors[1])
+                )
+            else:
+                raise ValueError("Context example pair must be either dense or sparse vectors")
 
-        # Edit query filter
-        ignore_mentioned_ids = models.HasIdCondition(has_id=mentioned_ids)
+        if sparse_context_vectors and dense_context_vectors:
+            raise ValueError("Context example pair must be either dense or sparse vectors")
```
Comment (coszio) [inline qdrant_client/local/local_collection.py:957]: Good check :)
```suggestion
            raise ValueError("All context example pairs must be either dense or sparse vectors")
```

**Review by coszio — APPROVED** [review]
Lots of work in a single PR, but still easy to go through. Nothing to complain aside from some nits

**Comment by joein** [comment]
tests are failing now due to #580

**Pull request merged**